fix(options_validator): add options if they do not exist

kvwalker opened the pull request:
While working on [NODE-1671](https://jira.mongodb.org/browse/NODE-1671), I realized that I was not handling the case where an operation is called without any options.

If `insertOne(doc, callback)` is called before this fix, the `doc` will be treated as `options` in `build()`. This fix ensures that a new options object is created if one is not passed in.

daprahamian commented on lib/options_validator.js:
```diff
@@ -209,9 +209,20 @@ class OperationBuilder {
-        typeof args[args.length - 1] === 'function' ? args.length - 2 : args.length - 1;
+      const hasCallback = typeof args[args.length - 1] === 'function';
+      let optionsIndex = hasCallback ? args.length - 2 : args.length - 1;
+      if (optionsIndex < operationBuilder.requiredArity) {
```
I'm not sure this is right. The algorithm we should be following for parsing parameters is basically:

```js
const callback = typeof args[args.length - 1] === 'function' ? args.pop() : whateverTheDefaultForCallbackIs;

if (args.length < arity || args.length > arity + 1) {
  // ERROR YOU HAVE THE WRONG # OF OPTIONS
}

const options = typeof args[arity] === object ? args.pop(): {};

args.push(options);
if (callback) {
  args.push(callback);
}
```

Basically, we would ensure the following:
+ We will assume a mistake in the following circumstances:
    + `args.length < arity`: Too few required args
    + `args.length > arity + 2`: Too many required args
    + `args.length === arity && typeof args[args.length - 1] === 'function'`: We assume the last arg is meant to be a callback, so there are too few required args
    + `args.length === arity + 2 && typeof args[args.length - 1] !== 'function'`: since last arg is not a callback, there are too many required args.
+ If someone passes `arity + 1` args, and the last arg is a callback, we will assume that they are omitting `options` and will provide a default
+ If someone passes `arity + 1` args, and the last arg is not a callback, we will assume that they are omitting `callback`, and that the last arg is `options`
+ If someone passes `arity + 2` args, and the last arg is a callback, we will assume that `args[arity]` is `options` and `args[arity + 1]` is `callback`

daprahamian commented on lib/options_validator.js:
```diff
@@ -209,23 +209,25 @@ class OperationBuilder {
-      const optionsIndex =
-        typeof args[args.length - 1] === 'function' ? args.length - 2 : args.length - 1;
+      const callback = typeof args[args.length - 1] === 'function' ? args.pop() : undefined;
+      const options = typeof args[operationBuilder.requiredArity] === 'object' ? args.pop(): {};
```
In this test case:

```js
const fn = arityOne().options({...}).build(...);
fn({a: 1}, {b: 2}, () => {}, callback);
```

This line will confirm `typeof {b: 2} === 'object'`, but then pop `() => {}` as the options object. I want to double check: will `assertArity` catch this, and warn before this happens?

mbroadst replied:
@kvwalker let's answer this with a unit test!

kvwalker commented:
I know you've both already approved this PR, but I found 2 more bugs that I fixed in the last commit. If you want, I can move this to a separate PR but I thought it would fit in here since it's another fix for options_validator.

The last commit fixes the following bugs:
- The [check](https://github.com/mongodb/node-mongodb-native/pull/1841/commits/597f70f6ceec2e7e2055c996d12afbd75c4c176d#diff-e1fa59f7e4e4c9080c24be96453dfdb8L96) for an option's default would behave incorrectly if the option's default was a `boolean`. I fixed it so it checks for the existence of a default with `hasOwnProperty`, rather than by getting the default value.
- Promises, like in [this test](https://github.com/mongodb/node-mongodb-native/blob/master/test/functional/cursor_tests.js#L4250), cause the `callback` to be `undefined`. Before [this fix](https://github.com/mongodb/node-mongodb-native/pull/1841/commits/597f70f6ceec2e7e2055c996d12afbd75c4c176d#diff-e1fa59f7e4e4c9080c24be96453dfdb8R31), `assertArity` would fail. Additionally, this undefined callback would cause [this](https://github.com/mongodb/node-mongodb-native/pull/1841/commits/597f70f6ceec2e7e2055c996d12afbd75c4c176d#diff-e1fa59f7e4e4c9080c24be96453dfdb8L212) to overwrite the callback by only pushing `options` onto `args`, so the aforementioned test would fail with `TypeError: final argument to `executeOperation` must be a callback`.

I wrote a unit test to make sure this fixes behavior with promises, but it's definitely not pretty so I'd welcome any suggestions to improve it.

daprahamian commented on test/unit/options_validator_tests.js:
```diff
@@ -494,4 +494,76 @@ describe('Options Validation', function() {
+    TestClass.prototype.testOperation = arityOne()
+      .options({ b: { type: 'boolean', default: false } })
+      .build(function(requiredArgument, options, callback) {
+        return this.testOperation2(requiredArgument, options, callback);
```
This is an interesting case, b/c ideally we do not want something to pass through options management more than once. Is this showing up somewhere in the actual code base?

kvwalker replied:
Yes, `insert` calls `insertMany`, which would cause the options passed to `insert` to go through options management twice.

daprahamian replied:
Ah. Ideally, we'd rework `insert` to call the `insertMany` operation instead. Still, I think it makes sense to add these checks.

kvwalker replied:
Yeah, that makes sense. I can make a ticket to rework `insert`.

mbroadst replied:
👍 to all of this 😄

mbroadst commented on test/unit/options_validator_tests.js:
```diff
@@ -445,4 +447,123 @@ describe('Options Validation', function() {
+    TestClass.prototype.testOperation2 = arityOne()
+      .options({ b: { type: 'boolean' } })
+      .build(function(requiredArgument, options, callback) {
+        return executeTestOperation(testOperation2, [this, requiredArgument, options, callback]);
```
can't you just use `executeOperation` from `utils.js` here, instead of implementing part of it in the test file?

kvwalker replied:
I was trying to avoid the `topology` and `session` logic inside of `executeOperation`.

mbroadst replied:
In that case, I think you're just trying to test a method that returns a `Promise` after executing right? Can't you just do:
```
function executeTestOperation(operation, args) {
  return new Promise(function(resolve, reject) {
    args[args.length - 1] = (err, result) => err ? reject(err) : resolve(result);;
    return operation.apply(null, args);
  });
}
```

mbroadst commented on lib/options_validator.js:
```diff
@@ -93,7 +95,10 @@ function validate(optionsSchema, providedOptions, overrideOptions, validationOpt
 
-      if (optionFromSchema.default && !verifiedOptions.hasOwnProperty(optionName)) {
+      if (
+        optionFromSchema.hasOwnProperty('default') &&
```
Is there a test for when `default` is a boolean?
